Runtime/Scripts/PersistentConnection.cs

```diff
@@ -296,104 +327,176 @@ IEnumerator OnIntervalEdgeEventsLatency(EdgeEventsConnection connection, string
+              return;
+            }
+            PropagateSuccess(FindCloudletEventTrigger.CloudletStateChanged, edgeEvent.NewCloudlet);
+          }
+          return;
+        case ServerEventType.EventCloudletUpdate:
+          if (fcTriggers.Contains(FindCloudletEventTrigger.CloserCloudlet))
+          {
+            if (edgeEvent.NewCloudlet == null)
+            {
+              PropagateError(FindCloudletEventTrigger.CloserCloudlet, "Received CloserCloudlet Event but NewCloudlet is null");
+              return;
+            }
+            PropagateSuccess(FindCloudletEventTrigger.CloserCloudlet, edgeEvent.NewCloudlet);
+          }
+          return;
+        case ServerEventType.EventError:
+          Logger.Log("Received EventError from server, " + edgeEvent.ToString());
+          PropagateError(FindCloudletEventTrigger.Error, edgeEvent.ErrorMsg);
+          return;
+        default:
+        case ServerEventType.EventUnknown:
+          Logger.Log("Received Unknown event, "+edgeEvent.ToString());
+          return;
       }
     }
 
-    async Task RespondToLatencyRequest()
+    IEnumerator RespondToLatencyRequest()
     {
       string host = integration.GetHost();
-      if (!LocationUpdate())
-      {
-        return;
-      }
+      yield return StartCoroutine(UpdateLocation());
       if (hasTCPPorts)
       {
         Logger.Log("Sending Latency Samples (TCP) as a response to LatencyRequest from the server");
-        await integration.matchingEngine.EdgeEventsConnection.TestConnectAndPostLatencyUpdate(host, (uint)config.latencyTestPort, location);
+        integration.matchingEngine.EdgeEventsConnection.TestConnectAndPostLatencyUpdate(host, (uint)config.latencyTestPort, location).ConfigureAwait(false);
       }
       else
       {
         Logger.Log("Sending Latency Samples (UDP) as a response to LatencyRequest from the server");
-        await integration.matchingEngine.EdgeEventsConnection.TestPingAndPostLatencyUpdate(host, location);
+        integration.matchingEngine.EdgeEventsConnection.TestPingAndPostLatencyUpdate(host, location).ConfigureAwait(false);
       }
     }
 
-    async void ProcessLatency(Statistics stats)
+    async Task ProcessLatency(Statistics stats)
     {
-      if (stats.Avg < config.latencyThresholdTriggerMs)
-      {
-        FindCloudletMode previousMode = integration.Mode;
-        integration.UseFindCloudletPerformanceMode(true);
-        bool fcResult = await integration.FindCloudlet();
-        if (fcResult)
+      Logger.Log("Processing Latency Received from Server");
+      if (stats.Avg > config.latencyThresholdTriggerMs)
+      {
+        Logger.Log("Current Latency (" + stats.Avg + ") > LatencyThreshold (" + config.latencyThresholdTriggerMs + ")");
+        Logger.Log("Performing FindCloudlet PerformanceMode");
+        MobiledgeXIntegration tempMxi = new MobiledgeXIntegration();
+        bool regResult = await tempMxi.Register("eu-qa.dme.mobiledgex.net", 50051);
+        if (regResult)
```

Comment:
nit: it's a little bit confusing to have a bunch of these nested if/elses. It might be easier to understand if you check for failures and return before moving on. ie.

```
if (!regResult)
{
    // propagate error and return
}
... find cloudlet code ...
if (!fcSuccess)
{
    // propagate error and return
}
... rest of code...
```

Reply:
we should return after propagating errors. (eg. we shouldn't be doing FindCloudlet if RegisterClient already failed)

Reply:
fixed